cpp/arcticdb/util/std_ranges_utils.hpp

```diff
@@ -0,0 +1,165 @@
+template<typename Container, typename KeyFunc>
+auto group_by(Container&& container, KeyFunc&& key_func) {
+    // Use decltype to get the correct value type from the range
+    using ValueType = std::remove_reference_t<decltype(*std::begin(container))>;
+    using KeyType = std::invoke_result_t<KeyFunc, ValueType>;
+    
+    std::unordered_map<KeyType, std::vector<ValueType>> result;
+    
+    for (auto&& item : std::forward<Container>(container)) {
+        auto key = std::invoke(std::forward<KeyFunc>(key_func), item);
+        result[std::move(key)].push_back(std::forward<ValueType>(item));
+    }
+    
+    return result;
+}
+
+// Standard library replacement for folly::gen::zip with move semantics
+template<typename Container1, typename Container2>
+auto zip_move(Container1&& container1, Container2&& container2) {
+    using ValueType1 = std::remove_reference_t<decltype(*std::begin(container1))>;
+    using ValueType2 = std::remove_reference_t<decltype(*std::begin(container2))>;
+    
+    std::vector<std::pair<ValueType1, ValueType2>> result;
+    result.reserve(std::min(container1.size(), container2.size()));
+    
+    auto it1 = std::begin(container1);
+    auto it2 = std::begin(container2);
+    auto end1 = std::end(container1);
+    auto end2 = std::end(container2);
+    
+    while (it1 != end1 && it2 != end2) {
+        result.emplace_back(std::move(*it1), std::move(*it2));
+        ++it1;
+        ++it2;
+    }
+    
+    return result;
+}
+
+// Helper for foreach on grouped data
+template<typename GroupedContainer, typename Func>
+void foreach_group(GroupedContainer&& grouped, Func&& func) {
+    for (auto&& [key, values] : std::forward<GroupedContainer>(grouped)) {
+        std::invoke(std::forward<Func>(func), key, std::forward<decltype(values)>(values));
+    }
+}
+
+// Standard library replacement for folly::gen::from with move semantics
+template<typename Container>
+auto from_move(Container&& container) {
+    return std::forward<Container>(container);
+}
+
+// Standard library replacement for folly::gen::map
+template<typename Container, typename Func>
+auto map(Container&& container, Func&& func) {
+    using InputType = std::remove_reference_t<decltype(*std::begin(container))>;
+    using OutputType = std::invoke_result_t<Func, InputType>;
+    
+    std::vector<OutputType> result;
+    result.reserve(container.size());
+    
+    for (auto&& item : std::forward<Container>(container)) {
+        result.push_back(std::invoke(std::forward<Func>(func), std::forward<InputType>(item)));
+    }
+    
+    return result;
+}
+
+// Standard library replacement for folly::gen::filter
+template<typename Container, typename Func>
+auto filter(Container&& container, Func&& func) {
+    using ValueType = std::remove_reference_t<decltype(*std::begin(container))>;
+    
+    std::vector<ValueType> result;
+    result.reserve(container.size()); // Reserve for worst case
+    
+    for (auto&& item : std::forward<Container>(container)) {
```

Comment:
Use std::ranges::copy_if with move iterators. I'd also pass std::back_inserter(result) to copy_if and don't bother with reserving the worst case and then calling shrink to fit.